stateresolution.go

```diff
@@ -0,0 +1,255 @@
+	return r.resolvedJoinRules, nil
+}
+
+func (r *stateResolver) ThirdPartyInvite(key string) (*Event, error) {
+	return r.resolvedThirdPartyInvites[key], nil
+}
+
+func (r *stateResolver) Member(key string) (*Event, error) {
+	return r.resolvedMembers[key], nil
+}
+
+func (r *stateResolver) addConflicted(events []Event) {
+	type conflictKey struct {
+		eventType string
+		stateKey  string
+	}
+	offsets := map[conflictKey]int{}
+	// Split up the conflicted events into blocks with the same state key.
+	for _, event := range events {
+		key := conflictKey{event.Type(), *event.StateKey()}
+		// Work out which block to add the event to.
+		var block *[]Event
+		// By default we add the event to a block in the others list.
+		blockList := &r.others
+		switch key.eventType {
+		case "m.room.create":
+			if key.stateKey == "" {
+				block = &r.creates
+			}
+		case "m.room.power_levels":
+			if key.stateKey == "" {
+				block = &r.powerLevels
+			}
+		case "m.room.join_rules":
+			if key.stateKey == "" {
+				block = &r.joinRules
+			}
+		case "m.room.member":
+			blockList = &r.members
+		case "m.room.third_party_invite":
+			blockList = &r.thirdPartyInvites
+		}
+		if block == nil {
+			// We need to find an entry for the state key in a block list.
+			offset, ok := offsets[key]
+			if !ok {
+				// This is the first time we've seen that state key so we add a
+				// new block to the block list.
+				offset = len(*blockList)
+				*blockList = append(*blockList, nil)
+			}
+			// Get the address of the block in the block list.
+			block = &(*blockList)[offset]
+		}
+		// Add the event to the block.
+		*block = append(*block, event)
+	}
+}
+
+// Add an event to the resolved auth events.
+func (r *stateResolver) addAuthEvent(event *Event) {
+	switch event.Type() {
+	case "m.room.create":
+		r.resolvedCreate = event
+	case "m.room.power_levels":
+		r.resolvedPowerLevels = event
+	case "m.room.join_rules":
+		r.resolvedJoinRules = event
+	case "m.room.member":
+		r.resolvedMembers[*event.StateKey()] = event
+	case "m.room.third_party_invite":
+		r.resolvedThirdPartyInvites[*event.StateKey()] = event
+	}
+}
+
+// Remove the auth event with the given type and state key.
+func (r *stateResolver) removeAuthEvent(eventType, stateKey string) {
+	switch eventType {
+	case "m.room.create":
+		r.resolvedCreate = nil
+	case "m.room.power_levels":
+		r.resolvedPowerLevels = nil
+	case "m.room.join_rules":
+		r.resolvedJoinRules = nil
+	case "m.room.member":
+		r.resolvedMembers[stateKey] = nil
+	case "m.room.third_party_invite":
+		r.resolvedThirdPartyInvites[stateKey] = nil
+	}
+}
+
+// resolveAndAddAuthBlocks resolves each block of conflicting auth state events in a list of blocks
+// where all the blocks have the same event type.
+// Once every block has been resolved the resulting events are added to the events used for auth checks.
+func (r *stateResolver) resolveAndAddAuthBlocks(blocks [][]Event) {
+	start := len(r.result)
```

Comment:
Given we blindly append to `result`, this makes these calls presumably unsafe to be called multiple times?